File: elements-of-programming-interviews/cpp/sorted_lists_merge.cc
#include "list_node.h"
#include "test_framework/generic_test.h"

// One / both could be empty
// Negatives?

shared_ptr<ListNode<int>> MergeTwoSortedLists(shared_ptr<ListNode<int>> L1,
                                              shared_ptr<ListNode<int>> L2) {
  shared_ptr<ListNode<int>> head = make_shared<ListNode<int>>();
  shared_ptr<ListNode<int>> curr = head;
  while (L1 && L2) {
    if (L1->data <= L2->data) {
      curr->next = L1;
      L1 = L1->next;
    } else {
      curr->next = L2;
      L2 = L2->next;
    }
    curr = curr->next;
  }
  if (L1) {
    curr->next = L1;
  }
  if (L2) {
    curr->next = L2;
  }

  return head->next;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L1", "L2"};
  return GenericTestMain(args, "sorted_lists_merge.cc",
                         "sorted_lists_merge.tsv", &MergeTwoSortedLists,
                         DefaultComparator{}, param_names);
}

File: elements-of-programming-interviews/2025/epi_judge_python/sorted_lists_merge.py
from typing import Optional

from list_node import ListNode
from test_framework import generic_test


def merge_two_sorted_lists(L1: Optional[ListNode],
                           L2: Optional[ListNode]) -> Optional[ListNode]:
    head = ListNode()
    curr = head
    while L1 and L2:
        if L1.data < L2.data:
            curr.next, L1 = L1, L1.next
        else:
            curr.next, L2 = L2, L2.next
        curr = curr.next 

    curr.next = L1 or L2 
    return head.next


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('sorted_lists_merge.py',
                                       'sorted_lists_merge.tsv',
                                       merge_two_sorted_lists))
